src/compiler/scala/tools/nsc/typechecker/Namers.scala

```diff
@@ -1675,6 +1696,9 @@ trait Namers extends MethodSynthesis {
 
   abstract class TypeCompleter extends LazyType {
     val tree: Tree
+    override def forceDirectSuperclasses: Unit = {
+      tree.foreach(t => Option(t.symbol).map(_.maybeInitialize))
```

Comment:
Agreed on the `DefTree` restriction.

On the sketchiness of `maybeInitialize` ... yes, side-effects won't be undone. However any side effects that are triggered before a cycle is observed should be stable, ie. we're merely moving them to an earlier point rather than changing the outcome.

Also, this path will only be followed if `knownDirectSubclasses` is called, and the result can't be any sketchier than the status quo.

Could we run a community build with this change and see how it plays out?